# Q1. Pandas Version

In [3]:
import pandas as pd

In [20]:
pd.__version__

'2.3.1'

## Getting the data

In [9]:
!pip install wget

The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.

[notice] A new release of pip is available: 25.1.1 -> 25.2
[notice] To update, run: python3 -m pip install --upgrade pip


In [11]:
!wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv

--2025-10-01 21:41:49--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.108.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 874188 (854K) [text/plain]
Saving to: ‘car_fuel_efficiency.csv’

car_fuel_efficiency 100%[===================>] 853.70K  --.-KB/s    in 0.02s   

2025-10-01 21:41:49 (35.4 MB/s) - ‘car_fuel_efficiency.csv’ saved [874188/874188]



In [16]:
df = pd.read_csv("car_fuel_efficiency.csv")

# Q2. Records Count, How many records are in the dataset?

In [18]:
df.shape

(9704, 11)

There are 9704 records.

# Q3. Fuel Types, How many fuel types are presented in the dataset?

In [25]:
list(df.columns)

['engine_displacement',
 'num_cylinders',
 'horsepower',
 'vehicle_weight',
 'acceleration',
 'model_year',
 'origin',
 'fuel_type',
 'drivetrain',
 'num_doors',
 'fuel_efficiency_mpg']

In [27]:
df['fuel_type'].unique()

array(['Gasoline', 'Diesel'], dtype=object)

There are 3 different fuel types in the dataset.

# Q4. Missing Values, How many columns in the dataset have missing values?

In [28]:
(df.isnull().sum() >0).sum()

np.int64(4)

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9704 entries, 0 to 9703
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   engine_displacement  9704 non-null   int64  
 1   num_cylinders        9222 non-null   float64
 2   horsepower           8996 non-null   float64
 3   vehicle_weight       9704 non-null   float64
 4   acceleration         8774 non-null   float64
 5   model_year           9704 non-null   int64  
 6   origin               9704 non-null   object 
 7   fuel_type            9704 non-null   object 
 8   drivetrain           9704 non-null   object 
 9   num_doors            9202 non-null   float64
 10  fuel_efficiency_mpg  9704 non-null   float64
dtypes: float64(6), int64(2), object(3)
memory usage: 834.1+ KB


There are 4 columns with missing values in the dataset.

# Q5. Max fuel efficiency, What's the maximum fuel efficiency of cars from Asia?

In [32]:
df.loc[df['origin'] == 'Asia', 'fuel_efficiency_mpg'].max()

np.float64(23.759122836520497)

The maximum fuel efficiency for Asia cars is 23.76 mpg.

# Q6. Median value horsepower:

    Find the median value of the horsepower column in the dataset.
    Next, calculate the most frequent value of the same horsepower column.
    Use the fillna method to fill the missing values in the horsepower column with the most frequent value from the previous step.
    Now, calculate the median value of horsepower once again.

Has it changed?

    Yes, it increased
    Yes, it decreased
    No


In [33]:
df['horsepower'].mean()

np.float64(149.65729212983547)

In [35]:
df['horsepower'].mode()[0]

np.float64(152.0)

In [42]:
df['horsepower'] = df['horsepower'].fillna(df['horsepower'].mode()[0])

In [43]:
df['horsepower'].mean()

np.float64(149.82821516900248)

Yes, it increased, just a little bit, but it increased.

# Q7. Sum of weights

    Select all the cars from Asia
    Select only columns vehicle_weight and model_year
    Select the first 7 values
    Get the underlying NumPy array. Let's call it X.
    Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
    Invert XTX.
    Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
    Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
    What's the sum of all the elements of the result?

    Note: You just implemented linear regression. We'll talk about it in the next lesson.

In [50]:
import numpy as np

In [44]:
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0       152.0     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

In [47]:
X = df.loc[df['origin'] == 'Asia', ['vehicle_weight', 'model_year']].head(7).to_numpy()

In [48]:
XTX = X.T @ X

In [51]:
XTX_inv = np.linalg.inv(XTX)

In [53]:
y = np.array([1100, 1300 , 800, 900, 1000, 1100, 1200])

In [54]:
w = XTX_inv @ X.T @ y

In [55]:
w.sum()

np.float64(0.5187709081074016)

The sum of all elements of the result is 0.51.